public/app/features/dashboard-scene/scene/setDashboardPanelContext.ts

```diff
@@ -0,0 +1,188 @@
+import { AnnotationChangeEvent, AnnotationEventUIModel, CoreApp, DataFrame } from '@grafana/data';
+import { AdHocFilterSet, dataLayers, SceneDataLayers, VizPanel } from '@grafana/scenes';
+import { DataSourceRef } from '@grafana/schema';
+import { AdHocFilterItem, PanelContext } from '@grafana/ui';
+import { deleteAnnotation, saveAnnotation, updateAnnotation } from 'app/features/annotations/api';
+
+import { getDashboardSceneFor, getPanelIdForVizPanel, getQueryRunnerFor } from '../utils/utils';
+
+import { DashboardScene } from './DashboardScene';
+
+export function setDashboardPanelContext(vizPanel: VizPanel, context: PanelContext) {
+  context.app = CoreApp.Dashboard;
+
+  context.canAddAnnotations = () => {
+    const dashboard = getDashboardSceneFor(vizPanel);
+    const builtInLayer = getBuiltInAnnotationsLayer(dashboard);
+
+    // When there is no builtin annotations query we disable the ability to add annotations
+    if (!builtInLayer || !dashboard.canEditDashboard()) {
+      return false;
+    }
+
+    // If RBAC is enabled there are additional conditions to check.
+    return Boolean(dashboard.state.meta.annotationsPermissions?.dashboard.canAdd);
+  };
+
+  context.canEditAnnotations = (dashboardUID?: string) => {
+    const dashboard = getDashboardSceneFor(vizPanel);
+
+    if (!dashboard.canEditDashboard()) {
+      return false;
+    }
+
+    if (dashboardUID) {
+      return Boolean(dashboard.state.meta.annotationsPermissions?.dashboard.canEdit);
+    }
+
+    return Boolean(dashboard.state.meta.annotationsPermissions?.organization.canEdit);
+  };
+
+  context.canDeleteAnnotations = (dashboardUID?: string) => {
+    const dashboard = getDashboardSceneFor(vizPanel);
+
+    if (!dashboard.canEditDashboard()) {
+      return false;
+    }
+
+    if (dashboardUID) {
+      return Boolean(dashboard.state.meta.annotationsPermissions?.dashboard.canDelete);
+    }
+
+    return Boolean(dashboard.state.meta.annotationsPermissions?.organization.canDelete);
+  };
+
+  context.onAnnotationCreate = async (event: AnnotationEventUIModel) => {
+    const dashboard = getDashboardSceneFor(vizPanel);
+
+    const isRegion = event.from !== event.to;
+    const anno = {
+      dashboardUID: dashboard.state.uid,
+      panelId: getPanelIdForVizPanel(vizPanel),
+      isRegion,
+      time: event.from,
+      timeEnd: isRegion ? event.to : 0,
+      tags: event.tags,
+      text: event.description,
+    };
+
+    await saveAnnotation(anno);
+
+    reRunBuiltInAnnotationsLayer(dashboard);
```

Comment:
This turns out really nice!

Reply:
@dprokop I agree! I was really happy you had added that function "runLayer" :heart: